zrml/prediction-markets/src/migrations.rs

```diff
@@ -15,7 +15,246 @@
+        total_weight
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<(), &'static str> {
+        Ok(())
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade() -> Result<(), &'static str> {
+        let dispute_duration = T::DisputePeriod::get();
+        let oracle_duration = T::ReportingPeriod::get();
+        let deadlines = Deadlines { grace_period: 0_u32.into(), oracle_duration, dispute_duration };
+        for (market_id, market) in storage_iter::<MarketOf<T>>(MARKET_COMMONS, MARKETS) {
+            assert_eq!(
+                market.deadlines, deadlines,
+                "found unexpected deadlines in market. market_id: {:?}, deadlines: {:?}",
+                market_id, market.deadlines
+            );
+        }
+        Ok(())
+    }
+}
+
+impl<T: Config> OnRuntimeUpgrade for MigrateMarketIdsPerBlockStorage<T> {
+    fn on_runtime_upgrade() -> Weight
+    where
+        T: Config,
+    {
+        let mut total_weight = T::DbWeight::get().reads(2);
+
+        if StorageVersion::get::<Pallet<T>>()
+            != PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION_FOR_MIGRATE_MARKET_IDS_STORAGE
+            || utility::get_on_chain_storage_version_of_market_commons_pallet()
+                != MARKET_COMMONS_NEXT_STORAGE_VERSION
+        {
+            log::info!(
+                "Skipping storage migration for MarketIds; prediction-markets already up to date"
+            );
+            return total_weight;
+        }
+
+        log::info!("Starting storage cleanup of MigrateMarketIdsPerBlockStorage");
+
+        type DisputeBlockToMarketIdsTuple<T> =
+            (<T as frame_system::Config>::BlockNumber, BoundedVec<MarketIdOf<T>, CacheSize>);
+        type IterType<T> = PrefixIterator<DisputeBlockToMarketIdsTuple<T>>;
+
+        let market_ids_per_dispute_iterator: IterType<T> =
+            frame_support::migration::storage_key_iter::<_, _, Twox64Concat>(
+                b"PredictionMarkets",
+                b"MarketIdsPerDisputeBlock",
+            );
+
+        let market_ids_per_dispute: Vec<DisputeBlockToMarketIdsTuple<T>> =
+            market_ids_per_dispute_iterator.collect();
+        total_weight = total_weight
+            .saturating_add(T::DbWeight::get().reads(market_ids_per_dispute.len() as u64));
+        for (dispute_start_block, market_ids) in &market_ids_per_dispute {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+            let dispute_duration = T::DisputePeriod::get();
+
+            let new_dispute_start_block = dispute_start_block.saturating_add(dispute_duration);
+            MarketIdsPerDisputeBlock::<T>::insert(new_dispute_start_block, market_ids);
+        }
+        for (dispute_start_block, _market_ids) in market_ids_per_dispute {
+            MarketIdsPerDisputeBlock::<T>::remove(dispute_start_block);
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+        }
```

Comment:
Don't you just come in conflict with these removes with what you have written above?

Reply:
Could you please elaborate? it can only be removing newly added data if dispute_duration is zero. which is not the case. right?

Reply:
You loop over `MarketIdsPerDisputeBlock ` two times. 
- In the first loop you set the default case for all previous markets, that the start block is now the future end block (after the dispute duration).  
- In the second for-loop you are removing all start blocks.

Problem with that: Conflicts for blocks, that are the same for future and current start blocks.
Example:
- A market ids of block 42 are moved (+8) with the first loop to block 50.
- B market ids of block 50 are moved (+8) with the first loop to block 58.

In the second for loop you would remove A's market ids completely.

Reply:
Good point. We noticed this problem earlier when the was just one loop and I just blurted out that splitting into two loops (one for removing, one for adding) would solve this problem.

I guess what you should really do is iterate in decreasing order. @vivekvpandya You should probably add a test case which ensures that this fixes @Chralt98's example.

Reply:
test added.